**The PR cannot be merged as the webapp fails** to start with the following error

> Caused by: java.lang.RuntimeException: Failure during SpringApplicationBuilder configuration: DSpace home directory could not be determined. It MUST include a subpath of '\config\config-definition.xml'. Please consider setting the 'dspace.dir' system property or ensure the dspace-api.jar is being run from [dspace]/lib/.
> 	at org.dspace.app.rest.Application.configure(Application.java:104)

The issue is mostly due to the dspace.home configuration property that look like as not yet injected in the dspaceHome attribute at the time that configure method is trigged. This is why now the DSpace kernel is instantiated "later" in the contextInitializer method https://github.com/DSpace/DSpace/blob/master/dspace-spring-rest/src/main/java/org/dspace/app/rest/Application.java#L84

If you use ENV variable to set the dspace.home you will not hit the issue but, obviously we cannot always force the use of this strategy as this will prevent to run multiple dspace instances on the same server

--- a/dspace-services/src/main/java/org/dspace/kernel/ServiceManager.java
+++ b/dspace-services/src/main/java/org/dspace/kernel/ServiceManager.java
@@ -7,6 +7,8 @@
  */
 package org.dspace.kernel;
 
+import org.springframework.context.ConfigurableApplicationContext;
+
 import java.util.List;
 import java.util.Map;
 
@@ -17,6 +19,11 @@ import java.util.Map;
  * @author Aaron Zeckoski (azeckoski @ gmail.com)
  */
 public interface ServiceManager {
+
+    /**
+     * Get the application context
+     */
+    public ConfigurableApplicationContext getApplicationContext();
 
     /**
      * Allows developers to get the desired service singleton by the provided type. <br>

--- a/dspace-services/src/main/java/org/dspace/kernel/ServiceManager.java
+++ b/dspace-services/src/main/java/org/dspace/kernel/ServiceManager.java
@@ -7,10 +7,10 @@
  */
 package org.dspace.kernel;
 
-import org.springframework.context.ConfigurableApplicationContext;
-
 import java.util.List;
 import java.util.Map;
+
+import org.springframework.context.ConfigurableApplicationContext;
 
 /**
  * Allows for non-specific access to the core services.

--- a/dspace-services/src/main/java/org/dspace/servicemanager/DSpaceServiceManager.java
+++ b/dspace-services/src/main/java/org/dspace/servicemanager/DSpaceServiceManager.java
@@ -7,16 +7,9 @@
  */
 package org.dspace.servicemanager;
 
-import java.util.*;
-import java.util.Map.Entry;
-
 import org.apache.commons.lang.ArrayUtils;
 import org.dspace.kernel.Activator;
-import org.dspace.kernel.mixins.ConfigChangeListener;
-import org.dspace.kernel.mixins.InitializedService;
-import org.dspace.kernel.mixins.ServiceChangeListener;
-import org.dspace.kernel.mixins.ServiceManagerReadyAware;
-import org.dspace.kernel.mixins.ShutdownService;
+import org.dspace.kernel.mixins.*;
 import org.dspace.servicemanager.config.DSpaceConfigurationService;
 import org.dspace.servicemanager.spring.SpringServiceManager;
 import org.dspace.services.ConfigurationService;
@@ -24,6 +17,10 @@ import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.BeanWrapper;
 import org.springframework.beans.PropertyAccessorFactory;
+import org.springframework.context.ConfigurableApplicationContext;
+
+import java.util.*;
+import java.util.Map.Entry;
 
 /**
  * This is the core service manager which ties together the other
@@ -299,6 +296,11 @@ public final class DSpaceServiceManager implements ServiceManagerSystem {
             }
         }
         return service;
+    }
+
+    @Override
+    public ConfigurableApplicationContext getApplicationContext() {
+        return primaryServiceManager.getApplicationContext();
     }
 
     public <T> List<T> getServicesByType(Class<T> type) {

--- a/dspace-services/src/main/java/org/dspace/servicemanager/DSpaceServiceManager.java
+++ b/dspace-services/src/main/java/org/dspace/servicemanager/DSpaceServiceManager.java
@@ -7,9 +7,22 @@
  */
 package org.dspace.servicemanager;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
 import org.apache.commons.lang.ArrayUtils;
 import org.dspace.kernel.Activator;
-import org.dspace.kernel.mixins.*;
+import org.dspace.kernel.mixins.ConfigChangeListener;
+import org.dspace.kernel.mixins.InitializedService;
+import org.dspace.kernel.mixins.ServiceChangeListener;
+import org.dspace.kernel.mixins.ServiceManagerReadyAware;
+import org.dspace.kernel.mixins.ShutdownService;
 import org.dspace.servicemanager.config.DSpaceConfigurationService;
 import org.dspace.servicemanager.spring.SpringServiceManager;
 import org.dspace.services.ConfigurationService;
@@ -18,9 +31,6 @@ import org.slf4j.LoggerFactory;
 import org.springframework.beans.BeanWrapper;
 import org.springframework.beans.PropertyAccessorFactory;
 import org.springframework.context.ConfigurableApplicationContext;
-
-import java.util.*;
-import java.util.Map.Entry;
 
 /**
  * This is the core service manager which ties together the other

--- a/dspace-services/src/main/java/org/dspace/servicemanager/spring/SpringServiceManager.java
+++ b/dspace-services/src/main/java/org/dspace/servicemanager/spring/SpringServiceManager.java
@@ -7,14 +7,6 @@
  */
 package org.dspace.servicemanager.spring;
 
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-
 import org.dspace.kernel.config.SpringLoader;
 import org.dspace.servicemanager.DSpaceServiceManager;
 import org.dspace.servicemanager.ServiceManagerSystem;
@@ -27,6 +19,8 @@ import org.springframework.beans.factory.NoSuchBeanDefinitionException;
 import org.springframework.beans.factory.config.AutowireCapableBeanFactory;
 import org.springframework.context.ApplicationContext;
 import org.springframework.context.support.ClassPathXmlApplicationContext;
+
+import java.util.*;
 
 /**
  * This is the Spring implementation of the service manager.
@@ -42,6 +36,7 @@ public final class SpringServiceManager implements ServiceManagerSystem {
     /**
      * @return the parent core Spring {@link ApplicationContext}
      */
+    @Override
     public ClassPathXmlApplicationContext getApplicationContext() {
         return applicationContext;
     }

--- a/dspace-services/src/main/java/org/dspace/servicemanager/spring/SpringServiceManager.java
+++ b/dspace-services/src/main/java/org/dspace/servicemanager/spring/SpringServiceManager.java
@@ -7,6 +7,14 @@
  */
 package org.dspace.servicemanager.spring;
 
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
 import org.dspace.kernel.config.SpringLoader;
 import org.dspace.servicemanager.DSpaceServiceManager;
 import org.dspace.servicemanager.ServiceManagerSystem;
@@ -19,8 +27,6 @@ import org.springframework.beans.factory.NoSuchBeanDefinitionException;
 import org.springframework.beans.factory.config.AutowireCapableBeanFactory;
 import org.springframework.context.ApplicationContext;
 import org.springframework.context.support.ClassPathXmlApplicationContext;
-
-import java.util.*;
 
 /**
  * This is the Spring implementation of the service manager.

--- a/dspace-services/src/test/java/org/dspace/servicemanager/MockServiceManagerSystem.java
+++ b/dspace-services/src/test/java/org/dspace/servicemanager/MockServiceManagerSystem.java
@@ -7,6 +7,8 @@
  */
 package org.dspace.servicemanager;
 
+import org.springframework.context.ConfigurableApplicationContext;
+
 import java.util.List;
 import java.util.Map;
 
@@ -59,9 +61,13 @@ public class MockServiceManagerSystem implements ServiceManagerSystem {
         return this.sms.getServiceByName(name, type);
     }
 
+    public ConfigurableApplicationContext getApplicationContext() {
+        return sms.getApplicationContext();
+    }
+
     /* (non-Javadoc)
-     * @see org.dspace.kernel.ServiceManager#getServicesByType(java.lang.Class)
-     */
+         * @see org.dspace.kernel.ServiceManager#getServicesByType(java.lang.Class)
+         */
     public <T> List<T> getServicesByType(Class<T> type) {
         return this.sms.getServicesByType(type);
     }

--- a/dspace-services/src/test/java/org/dspace/servicemanager/MockServiceManagerSystem.java
+++ b/dspace-services/src/test/java/org/dspace/servicemanager/MockServiceManagerSystem.java
@@ -7,10 +7,10 @@
  */
 package org.dspace.servicemanager;
 
-import org.springframework.context.ConfigurableApplicationContext;
-
 import java.util.List;
 import java.util.Map;
+
+import org.springframework.context.ConfigurableApplicationContext;
 
 
 /**

--- a/dspace-services/src/test/java/org/dspace/utils/servicemanager/ProviderStackTest.java
+++ b/dspace-services/src/test/java/org/dspace/utils/servicemanager/ProviderStackTest.java
@@ -7,16 +7,17 @@
  */
 package org.dspace.utils.servicemanager;
 
-import static org.junit.Assert.*;
+import org.dspace.kernel.ServiceManager;
+import org.dspace.kernel.mixins.OrderedService;
+import org.junit.Test;
+import org.springframework.context.ConfigurableApplicationContext;
 
 import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 
-import org.dspace.kernel.ServiceManager;
-import org.dspace.kernel.mixins.OrderedService;
-import org.junit.Test;
+import static org.junit.Assert.*;
 
 
 /**
@@ -60,6 +61,11 @@ public class ProviderStackTest {
             public <T> T getServiceByName(String name, Class<T> type) {
                 return null;
             }
+
+            public ConfigurableApplicationContext getApplicationContext() {
+                return null;
+            }
+
             public <T> List<T> getServicesByType(Class<T> type) {
                 return new ArrayList<T>();
             }

--- a/dspace-services/src/test/java/org/dspace/utils/servicemanager/ProviderStackTest.java
+++ b/dspace-services/src/test/java/org/dspace/utils/servicemanager/ProviderStackTest.java
@@ -7,18 +7,17 @@
  */
 package org.dspace.utils.servicemanager;
 
-import org.dspace.kernel.ServiceManager;
-import org.dspace.kernel.mixins.OrderedService;
-import org.junit.Test;
-import org.springframework.context.ConfigurableApplicationContext;
+import static org.junit.Assert.*;
 
 import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 
-import static org.junit.Assert.*;
-
+import org.dspace.kernel.ServiceManager;
+import org.dspace.kernel.mixins.OrderedService;
+import org.junit.Test;
+import org.springframework.context.ConfigurableApplicationContext;
 
 /**
  * Tests the usage of the provider stack

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/Application.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/Application.java
@@ -7,14 +7,6 @@
  */
 package org.dspace.app.rest;
 
-import java.io.File;
-
-import javax.naming.Context;
-import javax.naming.InitialContext;
-import javax.servlet.Filter;
-import javax.servlet.ServletContext;
-import javax.servlet.ServletException;
-
 import org.dspace.app.rest.filter.DSpaceRequestContextFilter;
 import org.dspace.app.rest.model.hateoas.DSpaceRelProvider;
 import org.dspace.app.rest.utils.ApplicationConfig;
@@ -26,21 +18,31 @@ import org.dspace.utils.servlet.DSpaceWebappServletFilter;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 import org.springframework.boot.builder.SpringApplicationBuilder;
 import org.springframework.boot.web.servlet.ServletContextInitializer;
 import org.springframework.boot.web.support.SpringBootServletInitializer;
+import org.springframework.context.ApplicationListener;
 import org.springframework.context.annotation.Bean;
+import org.springframework.context.event.ContextClosedEvent;
 import org.springframework.core.annotation.Order;
 import org.springframework.hateoas.RelProvider;
 import org.springframework.web.servlet.config.annotation.CorsRegistry;
 import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
 import org.springframework.web.servlet.config.annotation.WebMvcConfigurerAdapter;
 
+import javax.naming.Context;
+import javax.naming.InitialContext;
+import javax.servlet.Filter;
+import javax.servlet.ServletContext;
+import javax.servlet.ServletException;
+import java.io.File;
+
 /**
- * Define the Spring Boot Application settings itself. This class takes the place 
+ * Define the Spring Boot Application settings itself. This class takes the place
  * of a web.xml file, and configures all Filters/Listeners as methods (see below).
- * <P>
+ * <p>
  * NOTE: Requires a Servlet 3.0 container, e.g. Tomcat 7.0 or above.
  * <p>
  * NOTE: This extends SpringBootServletInitializer in order to allow us to build
@@ -51,92 +53,74 @@ import org.springframework.web.servlet.config.annotation.WebMvcConfigurerAdapter
  * @author Tim Donohue
  */
 @SpringBootApplication
-public class Application extends SpringBootServletInitializer
-{
+public class Application extends SpringBootServletInitializer {
     private static final Logger log = LoggerFactory.getLogger(Application.class);
 
     @Autowired
     private ApplicationConfig configuration;
-    
+
+    @Value("${dspace.dir}")
+    private String dspaceHome;
+
+    private transient DSpaceKernelImpl kernelImpl;
+
     /**
      * Override the default SpringBootServletInitializer.configure() method,
      * passing it this Application class.
-     * <P>
+     * <p>
      * This is necessary to allow us to build a deployable WAR, rather than
      * always relying on embedded Tomcat.
-     * <P>
+     * <p>
+     * <p>
      * See: http://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#howto-create-a-deployable-war-file
+     *
      * @param application
      * @return
      */
     @Override
     protected SpringApplicationBuilder configure(SpringApplicationBuilder application) {
-        return application.sources(Application.class);
+
+        // start the kernel when the webapp starts
+        try {
+            this.kernelImpl = DSpaceKernelInit.getKernel(null);
+            if (!this.kernelImpl.isRunning()) {
+                this.kernelImpl.start(getProvidedHome(dspaceHome)); // init the kernel
+            }
+
+            return application
+                    .parent(kernelImpl.getServiceManager().getApplicationContext())
+                    .listeners(new DSpaceKernelDestroyer(kernelImpl))
+                    .sources(Application.class);
+
+        } catch (Exception e) {
+            // failed to start so destroy it and log and throw an exception
+            try {
+                this.kernelImpl.destroy();
+            } catch (Exception e1) {
+                // nothing
+            }
+            String message = "Failure during SpringApplicationBuilder configuration: " + e.getMessage();
+            log.error(message + ":" + e.getMessage(), e);
+            throw new RuntimeException(message, e);
+        }
     }
 
     @Bean
     public ServletContextInitializer contextInitializer() {
         return new ServletContextInitializer() {
 
-        	private transient DSpaceKernelImpl kernelImpl;
-
             @Override
             public void onStartup(ServletContext servletContext)
                     throws ServletException {
-                    servletContext.setInitParameter("dspace.dir",configuration.getDspaceHome());
-                    
-                    // start the kernel when the webapp starts
-                    try {
-                        this.kernelImpl = DSpaceKernelInit.getKernel(null);
-                        if (! this.kernelImpl.isRunning()) {
-                        	this.kernelImpl.start(getProvidedHome(configuration.getDspaceHome())); // init the kernel
-                        }
-                    } catch (Exception e) {
-                        // failed to start so destroy it and log and throw an exception
-                        try {
-                            this.kernelImpl.destroy();
-                        } catch (Exception e1) {
-                            // nothing
-                        }
-                        String message = "Failure during filter init: " + e.getMessage();
-                        System.err.println(message + ":" + e);
-                        throw new RuntimeException(message, e);
-                    }
+                servletContext.setInitParameter("dspace.dir", configuration.getDspaceHome());
             }
-            
-            /*
-             * Find DSpace's "home" directory.
-             * Initially look for JNDI Resource called "java:/comp/env/dspace.dir".
-             * If not found, look for "dspace.dir" initial context parameter.
-             */
-            private String getProvidedHome(String dspaceHome){
-            	String providedHome = null;
-            	try {
-        			Context ctx = new InitialContext();
-        			providedHome = (String) ctx.lookup("java:/comp/env/" + DSpaceConfigurationService.DSPACE_HOME);
-        		} catch (Exception e) {
-        			// do nothing
-        		}
-        		
-        		if (providedHome == null)
-        		{
-        			if(dspaceHome != null && !dspaceHome.equals("") && 
-        					!dspaceHome.equals("${" + DSpaceConfigurationService.DSPACE_HOME + "}")){
-        				File test = new File(dspaceHome);
-        				if(test.exists() && new File(test,DSpaceConfigurationService.DSPACE_CONFIG_PATH).exists()) {
-        					providedHome = dspaceHome;
-                        }
-        			}
-        		}
-        		return providedHome;
-            }
-
         };
     }
 
     /**
      * Register the "DSpaceContextListener" so that it is loaded
      * for this Application.
+     *
      * @return DSpaceContextListener
      */
     @Bean
@@ -146,11 +130,11 @@ public class Application extends SpringBootServletInitializer
         // (and loads all DSpace configs)
         return new DSpaceContextListener();
     }
-    
+
     /**
      * Register the DSpaceWebappServletFilter, which initializes the
      * DSpace RequestService / SessionService
-     * 
+     *
      * @return DSpaceWebappServletFilter
      */
     @Bean
@@ -158,7 +142,7 @@ public class Application extends SpringBootServletInitializer
     protected Filter dspaceWebappServletFilter() {
         return new DSpaceWebappServletFilter();
     }
-    
+
     /**
      * Register the DSpaceRequestContextFilter, a Filter which checks for open
      * Context objects *after* a request has been fully processed, and closes them
@@ -170,12 +154,12 @@ public class Application extends SpringBootServletInitializer
     protected Filter dspaceRequestContextFilter() {
         return new DSpaceRequestContextFilter();
     }
-    
+
     @Bean
     protected RelProvider dspaceRelProvider() {
-    	return new DSpaceRelProvider();
+        return new DSpaceRelProvider();
     }
-    
+
     @Bean
     public WebMvcConfigurer corsConfigurer() {
         return new WebMvcConfigurerAdapter() {
@@ -183,9 +167,50 @@ public class Application extends SpringBootServletInitializer
             public void addCorsMappings(CorsRegistry registry) {
                 String[] corsAllowedOrigins = configuration.getCorsAllowedOrigins();
                 if (corsAllowedOrigins != null) {
-                	registry.addMapping("/api/**").allowedOrigins(corsAllowedOrigins);
+                    registry.addMapping("/api/**").allowedOrigins(corsAllowedOrigins);
                 }
             }
         };
+    }
+
+    /**
+     * Find DSpace's "home" directory.
+     * Initially look for JNDI Resource called "java:/comp/env/dspace.dir".
+     * If not found, look for "dspace.dir" initial context parameter.
+     */
+    private String getProvidedHome(String dspaceHome) {
+        String providedHome = null;
+        try {
+            Context ctx = new InitialContext();
+            providedHome = (String) ctx.lookup("java:/comp/env/" + DSpaceConfigurationService.DSPACE_HOME);
+        } catch (Exception e) {
+            // do nothing
+        }
+
+        if (providedHome == null) {
+            if (dspaceHome != null && !dspaceHome.equals("") &&
+                    !dspaceHome.equals("${" + DSpaceConfigurationService.DSPACE_HOME + "}")) {
+                File test = new File(dspaceHome);
+                if (test.exists() && new File(test, DSpaceConfigurationService.DSPACE_CONFIG_PATH).exists()) {
+                    providedHome = dspaceHome;
+                }
+            }
+        }
+        return providedHome;
+    }
+
+    private class DSpaceKernelDestroyer implements ApplicationListener<ContextClosedEvent> {
+        private DSpaceKernelImpl kernelImpl;
+
+        public DSpaceKernelDestroyer(DSpaceKernelImpl kernelImpl) {
+            this.kernelImpl = kernelImpl;
+        }
+
+        public void onApplicationEvent(final ContextClosedEvent event) {
+            if (this.kernelImpl != null) {
+                this.kernelImpl.destroy();
+                this.kernelImpl = null;
+            }
+        }
     }
 }

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/Application.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/Application.java
@@ -8,6 +8,7 @@
 package org.dspace.app.rest;
 
 import java.io.File;
+
 import javax.naming.Context;
 import javax.naming.InitialContext;
 import javax.servlet.Filter;

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/Application.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/Application.java
@@ -7,6 +7,13 @@
  */
 package org.dspace.app.rest;
 
+import java.io.File;
+import javax.naming.Context;
+import javax.naming.InitialContext;
+import javax.servlet.Filter;
+import javax.servlet.ServletContext;
+import javax.servlet.ServletException;
+
 import org.dspace.app.rest.filter.DSpaceRequestContextFilter;
 import org.dspace.app.rest.model.hateoas.DSpaceRelProvider;
 import org.dspace.app.rest.utils.ApplicationConfig;
@@ -18,12 +25,12 @@ import org.dspace.utils.servlet.DSpaceWebappServletFilter;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
-import org.springframework.beans.factory.annotation.Value;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 import org.springframework.boot.builder.SpringApplicationBuilder;
 import org.springframework.boot.web.servlet.ServletContextInitializer;
 import org.springframework.boot.web.support.SpringBootServletInitializer;
 import org.springframework.context.ApplicationListener;
+import org.springframework.context.ConfigurableApplicationContext;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.event.ContextClosedEvent;
 import org.springframework.core.annotation.Order;
@@ -31,13 +38,6 @@ import org.springframework.hateoas.RelProvider;
 import org.springframework.web.servlet.config.annotation.CorsRegistry;
 import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
 import org.springframework.web.servlet.config.annotation.WebMvcConfigurerAdapter;
-
-import javax.naming.Context;
-import javax.naming.InitialContext;
-import javax.servlet.Filter;
-import javax.servlet.ServletContext;
-import javax.servlet.ServletException;
-import java.io.File;
 
 /**
  * Define the Spring Boot Application settings itself. This class takes the place
@@ -54,15 +54,14 @@ import java.io.File;
  */
 @SpringBootApplication
 public class Application extends SpringBootServletInitializer {
+
     private static final Logger log = LoggerFactory.getLogger(Application.class);
 
     @Autowired
     private ApplicationConfig configuration;
 
-    @Value("${dspace.dir}")
-    private String dspaceHome;
-
-    private transient DSpaceKernelImpl kernelImpl;
+    @Autowired
+    private ConfigurableApplicationContext springBootApplicationContext;
 
     /**
      * Override the default SpringBootServletInitializer.configure() method,
@@ -79,41 +78,73 @@ public class Application extends SpringBootServletInitializer {
      */
     @Override
     protected SpringApplicationBuilder configure(SpringApplicationBuilder application) {
-
-        // start the kernel when the webapp starts
-        try {
-            this.kernelImpl = DSpaceKernelInit.getKernel(null);
-            if (!this.kernelImpl.isRunning()) {
-                this.kernelImpl.start(getProvidedHome(dspaceHome)); // init the kernel
-            }
-
-            return application
-                    .parent(kernelImpl.getServiceManager().getApplicationContext())
-                    .listeners(new DSpaceKernelDestroyer(kernelImpl))
-                    .sources(Application.class);
-
-        } catch (Exception e) {
-            // failed to start so destroy it and log and throw an exception
-            try {
-                this.kernelImpl.destroy();
-            } catch (Exception e1) {
-                // nothing
-            }
-            String message = "Failure during SpringApplicationBuilder configuration: " + e.getMessage();
-            log.error(message + ":" + e.getMessage(), e);
-            throw new RuntimeException(message, e);
-        }
+        return application.sources(Application.class);
     }
 
     @Bean
     public ServletContextInitializer contextInitializer() {
         return new ServletContextInitializer() {
 
+            private transient DSpaceKernelImpl kernelImpl;
+
             @Override
             public void onStartup(ServletContext servletContext)
                     throws ServletException {
                 servletContext.setInitParameter("dspace.dir", configuration.getDspaceHome());
+
+                // start the kernel when the webapp starts
+                try {
+                    this.kernelImpl = DSpaceKernelInit.getKernel(null);
+                    if (!this.kernelImpl.isRunning()) {
+                        this.kernelImpl.start(getProvidedHome(configuration.getDspaceHome())); // init the kernel
+                    }
+
+                    //Set the DSpace Kernel Application context as a parent of the Spring Boot context so that
+                    //we can auto-wire all DSpace Kernel services
+                    springBootApplicationContext.setParent(kernelImpl.getServiceManager().getApplicationContext());
+
+                    //Add a listener for Spring Boot application shutdown so that we can nicely cleanup the DSpace kernel.
+                    springBootApplicationContext.addApplicationListener(new DSpaceKernelDestroyer(kernelImpl));
+
+                } catch (Exception e) {
+                    // failed to start so destroy it and log and throw an exception
+                    try {
+                        this.kernelImpl.destroy();
+                    } catch (Exception e1) {
+                        // nothing
+                    }
+                    String message = "Failure during ServletContext initialisation: " + e.getMessage();
+                    log.error(message + ":" + e.getMessage(), e);
+                    throw new RuntimeException(message, e);
+                }
             }
+
+            /**
+             * Find DSpace's "home" directory.
+             * Initially look for JNDI Resource called "java:/comp/env/dspace.dir".
+             * If not found, look for "dspace.dir" initial context parameter.
+             */
+            private String getProvidedHome(String dspaceHome) {
+                String providedHome = null;
+                try {
+                    Context ctx = new InitialContext();
+                    providedHome = (String) ctx.lookup("java:/comp/env/" + DSpaceConfigurationService.DSPACE_HOME);
+                } catch (Exception e) {
+                    // do nothing
+                }
+
+                if (providedHome == null) {
+                    if (dspaceHome != null && !dspaceHome.equals("") &&
+                            !dspaceHome.equals("${" + DSpaceConfigurationService.DSPACE_HOME + "}")) {
+                        File test = new File(dspaceHome);
+                        if (test.exists() && new File(test, DSpaceConfigurationService.DSPACE_CONFIG_PATH).exists()) {
+                            providedHome = dspaceHome;
+                        }
+                    }
+                }
+                return providedHome;
+            }
+
         };
     }
 
@@ -173,32 +204,7 @@ public class Application extends SpringBootServletInitializer {
         };
     }
 
-    /**
-     * Find DSpace's "home" directory.
-     * Initially look for JNDI Resource called "java:/comp/env/dspace.dir".
-     * If not found, look for "dspace.dir" initial context parameter.
-     */
-    private String getProvidedHome(String dspaceHome) {
-        String providedHome = null;
-        try {
-            Context ctx = new InitialContext();
-            providedHome = (String) ctx.lookup("java:/comp/env/" + DSpaceConfigurationService.DSPACE_HOME);
-        } catch (Exception e) {
-            // do nothing
-        }
-
-        if (providedHome == null) {
-            if (dspaceHome != null && !dspaceHome.equals("") &&
-                    !dspaceHome.equals("${" + DSpaceConfigurationService.DSPACE_HOME + "}")) {
-                File test = new File(dspaceHome);
-                if (test.exists() && new File(test, DSpaceConfigurationService.DSPACE_CONFIG_PATH).exists()) {
-                    providedHome = dspaceHome;
-                }
-            }
-        }
-        return providedHome;
-    }
-
+    /** Utility class that will destory the DSpace Kernel on Spring Boot shutdown */
     private class DSpaceKernelDestroyer implements ApplicationListener<ContextClosedEvent> {
         private DSpaceKernelImpl kernelImpl;
 

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BitstreamFormatRestRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BitstreamFormatRestRepository.java
@@ -7,11 +7,13 @@
  */
 package org.dspace.app.rest.repository;
 
+import java.sql.SQLException;
+import java.util.List;
+
 import org.dspace.app.rest.converter.BitstreamFormatConverter;
 import org.dspace.app.rest.model.BitstreamFormatRest;
 import org.dspace.app.rest.model.hateoas.BitstreamFormatResource;
 import org.dspace.content.BitstreamFormat;
-import org.dspace.content.factory.ContentServiceFactory;
 import org.dspace.content.service.BitstreamFormatService;
 import org.dspace.core.Context;
 import org.springframework.beans.factory.annotation.Autowired;
@@ -19,8 +21,6 @@ import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Component;
 
-import java.sql.SQLException;
-import java.util.List;
 
 /**
  * This is the repository responsible to manage BitstreamFormat Rest object

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BitstreamFormatRestRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BitstreamFormatRestRepository.java
@@ -7,9 +7,6 @@
  */
 package org.dspace.app.rest.repository;
 
-import java.sql.SQLException;
-import java.util.List;
-
 import org.dspace.app.rest.converter.BitstreamFormatConverter;
 import org.dspace.app.rest.model.BitstreamFormatRest;
 import org.dspace.app.rest.model.hateoas.BitstreamFormatResource;
@@ -22,6 +19,9 @@ import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Component;
 
+import java.sql.SQLException;
+import java.util.List;
+
 /**
  * This is the repository responsible to manage BitstreamFormat Rest object
  * 
@@ -30,7 +30,10 @@ import org.springframework.stereotype.Component;
  */
 @Component(BitstreamFormatRest.CATEGORY + "." + BitstreamFormatRest.NAME)
 public class BitstreamFormatRestRepository extends DSpaceRestRepository<BitstreamFormatRest, Integer> {
-	BitstreamFormatService bfs = ContentServiceFactory.getInstance().getBitstreamFormatService();
+
+	@Autowired
+	BitstreamFormatService bfs;
+
 	@Autowired
 	BitstreamFormatConverter converter;
 

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BitstreamRestRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BitstreamRestRepository.java
@@ -7,20 +7,11 @@
  */
 package org.dspace.app.rest.repository;
 
-import java.io.IOException;
-import java.io.InputStream;
-import java.sql.SQLException;
-import java.util.ArrayList;
-import java.util.Iterator;
-import java.util.List;
-import java.util.UUID;
-
 import org.dspace.app.rest.converter.BitstreamConverter;
 import org.dspace.app.rest.model.BitstreamRest;
 import org.dspace.app.rest.model.hateoas.BitstreamResource;
 import org.dspace.authorize.AuthorizeException;
 import org.dspace.content.Bitstream;
-import org.dspace.content.factory.ContentServiceFactory;
 import org.dspace.content.service.BitstreamService;
 import org.dspace.core.Context;
 import org.springframework.beans.factory.annotation.Autowired;
@@ -28,6 +19,14 @@ import org.springframework.data.domain.Page;
 import org.springframework.data.domain.PageImpl;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.UUID;
 
 /**
  * This is the repository responsible to manage Bitstream Rest object
@@ -38,7 +37,10 @@ import org.springframework.stereotype.Component;
 
 @Component(BitstreamRest.CATEGORY + "." + BitstreamRest.NAME)
 public class BitstreamRestRepository extends DSpaceRestRepository<BitstreamRest, UUID> {
-	BitstreamService bs = ContentServiceFactory.getInstance().getBitstreamService();
+
+	@Autowired
+	BitstreamService bs;
+
 	@Autowired
 	BitstreamConverter converter;
 	

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BitstreamRestRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BitstreamRestRepository.java
@@ -7,6 +7,14 @@
  */
 package org.dspace.app.rest.repository;
 
+import java.io.IOException;
+import java.io.InputStream;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.UUID;
+
 import org.dspace.app.rest.converter.BitstreamConverter;
 import org.dspace.app.rest.model.BitstreamRest;
 import org.dspace.app.rest.model.hateoas.BitstreamResource;
@@ -19,14 +27,6 @@ import org.springframework.data.domain.Page;
 import org.springframework.data.domain.PageImpl;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Component;
-
-import java.io.IOException;
-import java.io.InputStream;
-import java.sql.SQLException;
-import java.util.ArrayList;
-import java.util.Iterator;
-import java.util.List;
-import java.util.UUID;
 
 /**
  * This is the repository responsible to manage Bitstream Rest object

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BrowseEntryLinkRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BrowseEntryLinkRepository.java
@@ -7,38 +7,28 @@
  */
 package org.dspace.app.rest.repository;
 
-import java.sql.SQLException;
-import java.util.Arrays;
-import java.util.Iterator;
-import java.util.UUID;
-import java.util.function.Consumer;
-
-import javax.servlet.http.HttpServletRequest;
-
 import org.apache.commons.lang3.StringUtils;
 import org.dspace.app.rest.converter.BrowseEntryConverter;
 import org.dspace.app.rest.converter.BrowseIndexConverter;
 import org.dspace.app.rest.model.BrowseEntryRest;
 import org.dspace.app.rest.model.BrowseIndexRest;
 import org.dspace.app.rest.model.hateoas.BrowseEntryResource;
-import org.dspace.browse.BrowseEngine;
-import org.dspace.browse.BrowseException;
-import org.dspace.browse.BrowseIndex;
-import org.dspace.browse.BrowseInfo;
-import org.dspace.browse.BrowserScope;
+import org.dspace.browse.*;
 import org.dspace.content.DSpaceObject;
-import org.dspace.content.factory.ContentServiceFactory;
 import org.dspace.content.service.CollectionService;
 import org.dspace.content.service.CommunityService;
 import org.dspace.core.Context;
 import org.springframework.beans.factory.annotation.Autowired;
-import org.springframework.data.domain.Page;
-import org.springframework.data.domain.PageImpl;
-import org.springframework.data.domain.PageRequest;
-import org.springframework.data.domain.Pageable;
-import org.springframework.data.domain.Sort;
+import org.springframework.data.domain.*;
 import org.springframework.data.domain.Sort.Order;
 import org.springframework.stereotype.Component;
+
+import javax.servlet.http.HttpServletRequest;
+import java.sql.SQLException;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.UUID;
+import java.util.function.Consumer;
 
 /**
  * This is the repository responsible to retrieve the first level values
@@ -56,9 +46,11 @@ public class BrowseEntryLinkRepository extends AbstractDSpaceRestRepository
 	@Autowired
 	BrowseIndexConverter bixConverter;
 
-	CollectionService collectionService = ContentServiceFactory.getInstance().getCollectionService();
+	@Autowired
+	CollectionService collectionService;
 
-	CommunityService communityService = ContentServiceFactory.getInstance().getCommunityService();
+	@Autowired
+	CommunityService communityService;
 
 	// FIXME It will be nice to drive arguments binding by annotation as in normal spring controller methods
 	public Page<BrowseEntryRest> listBrowseEntries(HttpServletRequest request, String browseName,

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BrowseEntryLinkRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BrowseEntryLinkRepository.java
@@ -7,28 +7,36 @@
  */
 package org.dspace.app.rest.repository;
 
+import java.sql.SQLException;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.UUID;
+import java.util.function.Consumer;
+import javax.servlet.http.HttpServletRequest;
+
 import org.apache.commons.lang3.StringUtils;
 import org.dspace.app.rest.converter.BrowseEntryConverter;
 import org.dspace.app.rest.converter.BrowseIndexConverter;
 import org.dspace.app.rest.model.BrowseEntryRest;
 import org.dspace.app.rest.model.BrowseIndexRest;
 import org.dspace.app.rest.model.hateoas.BrowseEntryResource;
-import org.dspace.browse.*;
+import org.dspace.browse.BrowseEngine;
+import org.dspace.browse.BrowseException;
+import org.dspace.browse.BrowseIndex;
+import org.dspace.browse.BrowseInfo;
+import org.dspace.browse.BrowserScope;
 import org.dspace.content.DSpaceObject;
 import org.dspace.content.service.CollectionService;
 import org.dspace.content.service.CommunityService;
 import org.dspace.core.Context;
 import org.springframework.beans.factory.annotation.Autowired;
-import org.springframework.data.domain.*;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Sort;
 import org.springframework.data.domain.Sort.Order;
 import org.springframework.stereotype.Component;
-
-import javax.servlet.http.HttpServletRequest;
-import java.sql.SQLException;
-import java.util.Arrays;
-import java.util.Iterator;
-import java.util.UUID;
-import java.util.function.Consumer;
 
 /**
  * This is the repository responsible to retrieve the first level values
@@ -59,7 +67,7 @@ public class BrowseEntryLinkRepository extends AbstractDSpaceRestRepository
 		// argument
 		String scope = null;
 		if (request != null) {
-			request.getParameter("scope");
+			scope = request.getParameter("scope");
 		}
 
 		Context context = obtainContext();

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BrowseItemLinkRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BrowseItemLinkRepository.java
@@ -10,7 +10,6 @@ package org.dspace.app.rest.repository;
 import java.sql.SQLException;
 import java.util.Iterator;
 import java.util.UUID;
-
 import javax.servlet.http.HttpServletRequest;
 
 import org.apache.commons.lang3.StringUtils;
@@ -25,7 +24,6 @@ import org.dspace.browse.BrowseInfo;
 import org.dspace.browse.BrowserScope;
 import org.dspace.content.DSpaceObject;
 import org.dspace.content.Item;
-import org.dspace.content.factory.ContentServiceFactory;
 import org.dspace.content.service.CollectionService;
 import org.dspace.content.service.CommunityService;
 import org.dspace.core.Context;

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BrowseItemLinkRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/BrowseItemLinkRepository.java
@@ -56,9 +56,11 @@ public class BrowseItemLinkRepository extends AbstractDSpaceRestRepository
 	@Autowired
 	ItemRestRepository itemRestRepository;
 
-	CollectionService collectionService = ContentServiceFactory.getInstance().getCollectionService();
-	
-	CommunityService communityService = ContentServiceFactory.getInstance().getCommunityService();
+	@Autowired
+	CollectionService collectionService;
+
+	@Autowired
+	CommunityService communityService;
 
 	public Page<ItemRest> listBrowseItems(HttpServletRequest request, String browseName, Pageable pageable, String projection)
 			throws BrowseException, SQLException {

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/CollectionRestRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/CollectionRestRepository.java
@@ -7,11 +7,6 @@
  */
 package org.dspace.app.rest.repository;
 
-import java.sql.SQLException;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.UUID;
-
 import org.dspace.app.rest.converter.CollectionConverter;
 import org.dspace.app.rest.model.CollectionRest;
 import org.dspace.app.rest.model.hateoas.CollectionResource;
@@ -25,6 +20,11 @@ import org.springframework.data.domain.PageImpl;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Component;
 
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
 /**
  * This is the repository responsible to manage Item Rest object
  * 
@@ -34,7 +34,10 @@ import org.springframework.stereotype.Component;
 
 @Component(CollectionRest.CATEGORY + "." + CollectionRest.NAME)
 public class CollectionRestRepository extends DSpaceRestRepository<CollectionRest, UUID> {
-	CollectionService cs = ContentServiceFactory.getInstance().getCollectionService();
+
+	@Autowired
+	CollectionService cs;
+
 	@Autowired
 	CollectionConverter converter;
 	

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/CollectionRestRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/CollectionRestRepository.java
@@ -7,11 +7,15 @@
  */
 package org.dspace.app.rest.repository;
 
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
 import org.dspace.app.rest.converter.CollectionConverter;
 import org.dspace.app.rest.model.CollectionRest;
 import org.dspace.app.rest.model.hateoas.CollectionResource;
 import org.dspace.content.Collection;
-import org.dspace.content.factory.ContentServiceFactory;
 import org.dspace.content.service.CollectionService;
 import org.dspace.core.Context;
 import org.springframework.beans.factory.annotation.Autowired;
@@ -19,11 +23,6 @@ import org.springframework.data.domain.Page;
 import org.springframework.data.domain.PageImpl;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Component;
-
-import java.sql.SQLException;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.UUID;
 
 /**
  * This is the repository responsible to manage Item Rest object

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java
@@ -7,11 +7,6 @@
  */
 package org.dspace.app.rest.repository;
 
-import java.sql.SQLException;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.UUID;
-
 import org.dspace.app.rest.converter.CommunityConverter;
 import org.dspace.app.rest.model.CommunityRest;
 import org.dspace.app.rest.model.hateoas.CommunityResource;
@@ -25,6 +20,11 @@ import org.springframework.data.domain.PageImpl;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Component;
 
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
 /**
  * This is the repository responsible to manage Item Rest object
  * 
@@ -34,7 +34,10 @@ import org.springframework.stereotype.Component;
 
 @Component(CommunityRest.CATEGORY + "." + CommunityRest.NAME)
 public class CommunityRestRepository extends DSpaceRestRepository<CommunityRest, UUID> {
-	CommunityService cs = ContentServiceFactory.getInstance().getCommunityService();
+
+	@Autowired
+	CommunityService cs;
+
 	@Autowired
 	CommunityConverter converter;
 	

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java
@@ -7,11 +7,15 @@
  */
 package org.dspace.app.rest.repository;
 
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
 import org.dspace.app.rest.converter.CommunityConverter;
 import org.dspace.app.rest.model.CommunityRest;
 import org.dspace.app.rest.model.hateoas.CommunityResource;
 import org.dspace.content.Community;
-import org.dspace.content.factory.ContentServiceFactory;
 import org.dspace.content.service.CommunityService;
 import org.dspace.core.Context;
 import org.springframework.beans.factory.annotation.Autowired;
@@ -19,11 +23,6 @@ import org.springframework.data.domain.Page;
 import org.springframework.data.domain.PageImpl;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Component;
-
-import java.sql.SQLException;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.UUID;
 
 /**
  * This is the repository responsible to manage Item Rest object

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java
@@ -7,12 +7,6 @@
  */
 package org.dspace.app.rest.repository;
 
-import java.sql.SQLException;
-import java.util.ArrayList;
-import java.util.Iterator;
-import java.util.List;
-import java.util.UUID;
-
 import org.dspace.app.rest.converter.ItemConverter;
 import org.dspace.app.rest.model.ItemRest;
 import org.dspace.app.rest.model.hateoas.ItemResource;
@@ -26,6 +20,12 @@ import org.springframework.data.domain.PageImpl;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Component;
 
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.UUID;
+
 /**
  * This is the repository responsible to manage Item Rest object
  * 
@@ -35,7 +35,10 @@ import org.springframework.stereotype.Component;
 
 @Component(ItemRest.CATEGORY + "." + ItemRest.NAME)
 public class ItemRestRepository extends DSpaceRestRepository<ItemRest, UUID> {
-	ItemService is = ContentServiceFactory.getInstance().getItemService();
+
+	@Autowired
+	ItemService is;
+
 	@Autowired
 	ItemConverter converter;
 	

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java
@@ -7,11 +7,16 @@
  */
 package org.dspace.app.rest.repository;
 
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.UUID;
+
 import org.dspace.app.rest.converter.ItemConverter;
 import org.dspace.app.rest.model.ItemRest;
 import org.dspace.app.rest.model.hateoas.ItemResource;
 import org.dspace.content.Item;
-import org.dspace.content.factory.ContentServiceFactory;
 import org.dspace.content.service.ItemService;
 import org.dspace.core.Context;
 import org.springframework.beans.factory.annotation.Autowired;
@@ -19,12 +24,6 @@ import org.springframework.data.domain.Page;
 import org.springframework.data.domain.PageImpl;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Component;
-
-import java.sql.SQLException;
-import java.util.ArrayList;
-import java.util.Iterator;
-import java.util.List;
-import java.util.UUID;
 
 /**
  * This is the repository responsible to manage Item Rest object

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/MetadataFieldRestRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/MetadataFieldRestRepository.java
@@ -7,20 +7,19 @@
  */
 package org.dspace.app.rest.repository;
 
+import java.sql.SQLException;
+import java.util.List;
+
 import org.dspace.app.rest.converter.MetadataFieldConverter;
 import org.dspace.app.rest.model.MetadataFieldRest;
 import org.dspace.app.rest.model.hateoas.MetadataFieldResource;
 import org.dspace.content.MetadataField;
-import org.dspace.content.factory.ContentServiceFactory;
 import org.dspace.content.service.MetadataFieldService;
 import org.dspace.core.Context;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Component;
-
-import java.sql.SQLException;
-import java.util.List;
 
 /**
  * This is the repository responsible to manage MetadataField Rest object

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/MetadataFieldRestRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/MetadataFieldRestRepository.java
@@ -7,9 +7,6 @@
  */
 package org.dspace.app.rest.repository;
 
-import java.sql.SQLException;
-import java.util.List;
-
 import org.dspace.app.rest.converter.MetadataFieldConverter;
 import org.dspace.app.rest.model.MetadataFieldRest;
 import org.dspace.app.rest.model.hateoas.MetadataFieldResource;
@@ -22,6 +19,9 @@ import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Component;
 
+import java.sql.SQLException;
+import java.util.List;
+
 /**
  * This is the repository responsible to manage MetadataField Rest object
  * 
@@ -30,7 +30,10 @@ import org.springframework.stereotype.Component;
  */
 @Component(MetadataFieldRest.CATEGORY + "." + MetadataFieldRest.NAME)
 public class MetadataFieldRestRepository extends DSpaceRestRepository<MetadataFieldRest, Integer> {
-	MetadataFieldService metaFieldService = ContentServiceFactory.getInstance().getMetadataFieldService();
+
+	@Autowired
+	MetadataFieldService metaFieldService;
+
 	@Autowired
 	MetadataFieldConverter converter;
 

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/MetadataSchemaRestRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/MetadataSchemaRestRepository.java
@@ -7,20 +7,19 @@
  */
 package org.dspace.app.rest.repository;
 
+import java.sql.SQLException;
+import java.util.List;
+
 import org.dspace.app.rest.converter.MetadataSchemaConverter;
 import org.dspace.app.rest.model.MetadataSchemaRest;
 import org.dspace.app.rest.model.hateoas.MetadataSchemaResource;
 import org.dspace.content.MetadataSchema;
-import org.dspace.content.factory.ContentServiceFactory;
 import org.dspace.content.service.MetadataSchemaService;
 import org.dspace.core.Context;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Component;
-
-import java.sql.SQLException;
-import java.util.List;
 
 /**
  * This is the repository responsible to manage MetadataSchema Rest object

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/MetadataSchemaRestRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/MetadataSchemaRestRepository.java
@@ -7,9 +7,6 @@
  */
 package org.dspace.app.rest.repository;
 
-import java.sql.SQLException;
-import java.util.List;
-
 import org.dspace.app.rest.converter.MetadataSchemaConverter;
 import org.dspace.app.rest.model.MetadataSchemaRest;
 import org.dspace.app.rest.model.hateoas.MetadataSchemaResource;
@@ -22,6 +19,9 @@ import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Component;
 
+import java.sql.SQLException;
+import java.util.List;
+
 /**
  * This is the repository responsible to manage MetadataSchema Rest object
  * 
@@ -30,7 +30,10 @@ import org.springframework.stereotype.Component;
  */
 @Component(MetadataSchemaRest.CATEGORY + "." + MetadataSchemaRest.NAME)
 public class MetadataSchemaRestRepository extends DSpaceRestRepository<MetadataSchemaRest, Integer> {
-	MetadataSchemaService metaScemaService = ContentServiceFactory.getInstance().getMetadataSchemaService();
+
+	@Autowired
+	MetadataSchemaService metaScemaService;
+
 	@Autowired
 	MetadataSchemaConverter converter;
 

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/SiteRestRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/SiteRestRepository.java
@@ -7,11 +7,15 @@
  */
 package org.dspace.app.rest.repository;
 
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
 import org.dspace.app.rest.converter.SiteConverter;
 import org.dspace.app.rest.model.SiteRest;
 import org.dspace.app.rest.model.hateoas.SiteResource;
 import org.dspace.content.Site;
-import org.dspace.content.factory.ContentServiceFactory;
 import org.dspace.content.service.SiteService;
 import org.dspace.core.Context;
 import org.springframework.beans.factory.annotation.Autowired;
@@ -19,11 +23,6 @@ import org.springframework.data.domain.Page;
 import org.springframework.data.domain.PageImpl;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Component;
-
-import java.sql.SQLException;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.UUID;
 
 /**
  * This is the repository responsible to manage Item Rest object

--- a/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/SiteRestRepository.java
+++ b/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/SiteRestRepository.java
@@ -7,11 +7,6 @@
  */
 package org.dspace.app.rest.repository;
 
-import java.sql.SQLException;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.UUID;
-
 import org.dspace.app.rest.converter.SiteConverter;
 import org.dspace.app.rest.model.SiteRest;
 import org.dspace.app.rest.model.hateoas.SiteResource;
@@ -25,6 +20,11 @@ import org.springframework.data.domain.PageImpl;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Component;
 
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
 /**
  * This is the repository responsible to manage Item Rest object
  * 
@@ -34,7 +34,10 @@ import org.springframework.stereotype.Component;
 
 @Component(SiteRest.CATEGORY + "." + SiteRest.NAME)
 public class SiteRestRepository extends DSpaceRestRepository<SiteRest, UUID> {
-	SiteService sitesv = ContentServiceFactory.getInstance().getSiteService();
+
+	@Autowired
+	SiteService sitesv;
+
 	@Autowired
 	SiteConverter converter;
 	

